chia/simulator/setup_nodes.py

```diff
@@ -409,103 +358,99 @@ async def setup_full_system_inner(
+                async with setup_harvester(
+                    shared_b_tools,
+                    shared_b_tools.root_path / "harvester",
+                    UnresolvedPeerInfo(shared_b_tools.config["self_hostname"], farmer_service._server.get_port()),
+                    consensus_constants,
+                ) as harvester_service:
+                    harvester = harvester_service._node
+
+                    vdf1_port = uint16(find_available_listen_port("vdf1"))
+                    vdf2_port = uint16(find_available_listen_port("vdf2"))
+                    async with setup_timelord(
+                        full_node_0_port,
+                        False,
+                        consensus_constants,
+                        b_tools.config,
+                        b_tools.root_path,
+                        vdf_port=vdf1_port,
+                    ) as timelord:
+                        async with setup_timelord(
+                            uint16(1000),
+                            True,
+                            consensus_constants,
+                            b_tools_1.config,
+                            b_tools_1.root_path,
+                            vdf_port=vdf2_port,
+                        ) as timelord_bluebox_service:
+
+                            async def num_connections() -> int:
+                                count = len(harvester.server.all_connections.items())
+                                return count
+
+                            await time_out_assert_custom_interval(10, 3, num_connections, 1)
+                            async with setup_vdf_clients(
+                                shared_b_tools, shared_b_tools.config["self_hostname"], vdf1_port
+                            ) as vdf_clients:
+                                async with setup_vdf_client(
+                                    shared_b_tools, shared_b_tools.config["self_hostname"], vdf2_port
+                                ) as vdf_bluebox_clients:
+                                    timelord_bluebox = timelord_bluebox_service._api
+                                    timelord_bluebox_server = timelord_bluebox_service._node.server
+                                    ret = (
+                                        node_apis[0],
+                                        node_apis[1],
+                                        harvester,
+                                        farmer_service._node,
+                                        introducer,
+                                        timelord,
+                                        vdf_clients,
+                                        vdf_bluebox_clients,
+                                        timelord_bluebox,
+                                        timelord_bluebox_server,
+                                    )
+                                    async with setup_daemon(btools=b_tools) as daemon_ws:
+                                        yield daemon_ws, ret
```

Comment:
it seems to me that you may be able to avoid this deep nesting with the exit stack, no?

Reply:
In general, I wouldn't suggest using an exit stack just to eliminate nesting.  In this case...  maybe it's special enough?  They all exit at the same time...  It's really deep...

Reply:
Yeah I agree with Arvid, I think the intention here if written again fresh would have been to just have them all part of one context

Reply:
The nesting has been eliminated